datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,336 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Distribution partitioner class that works with Hugging Face Datasets."""
+
+
+from collections import Counter
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class DistributionPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on a distribution.
+
+    Inspired from implementations of Li et al.
+    Federated Optimization in Heterogeneous Networks (2020)
+    https://arxiv.org/abs/1812.06127.
+
+    Parameters
+    ----------
+    distribution_array : Union[NDArrayInt, NDArrayFloat]
+        A 2-dimensional numpy array of the probability distribution of samples
+        for all labels in all partitions. The array shape should be
+        (`num_unique_labels`,
+        `num_unique_labels_per_partition*num_partitions/num_unique_labels`),
+        such that the first row of the array corresponds to the sample distribution
+        of the first unique label (in ascending order). The values may be scaled per
+        label such that the sum of the label distributions across all partitions are
+        equal (or close to) to the original unpartitioned label distribution
+        - see the `rescale` argument.
+    num_partitions : int
+        The total number of partitions that the data will be divided into.
+    num_unique_labels_per_partition : int
```

Comment:
This number needs to be checked too. Currently, it works also when the `num_unique_labels_per_client` is bigger than the number of `unique_labels` e.g.  `num_unique_labels_per_client = 20` in the example that you provided.